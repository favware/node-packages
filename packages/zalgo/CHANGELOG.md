# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [2.0.0](https://github.com/favware/node-packages/compare/@favware/zalgo@2.0.0-canary.0...@favware/zalgo@2.0.0) (2020-05-06)

**Note:** Version bump only for package @favware/zalgo





# [2.0.0-canary.0](https://github.com/favware/node-packages/compare/@favware/zalgo@1.1.5...@favware/zalgo@2.0.0-canary.0) (2020-05-06)


### Bug Fixes

* reworked to compile with tsc rather than rollup ([0b7c6c8](https://github.com/favware/node-packages/commit/0b7c6c81fab75fd298eea8427bbee373d91306bb))


### BREAKING CHANGES

* The default target for ts-config and these packages is now CommonJS, which will
work in NodeJS by default and get resolved in Rollup (with @rollup/plugin-commonjs) and Webpack as
well. Furthermore no longer will packages be compiled with importHelpers so tslib is no longer a
hidden required dependency.





## [1.1.5](https://github.com/favware/node-packages/compare/@favware/zalgo@1.1.4...@favware/zalgo@1.1.5) (2020-04-01)

**Note:** Version bump only for package @favware/zalgo

## [1.1.4](https://github.com/favware/node-packages/compare/@favware/zalgo@1.1.3...@favware/zalgo@1.1.4) (2020-01-22)

**Note:** Version bump only for package @favware/zalgo

## [1.1.3](https://github.com/favware/node-packages/compare/@favware/zalgo@1.1.2...@favware/zalgo@1.1.3) (2019-11-09)

**Note:** Version bump only for package @favware/zalgo

## [1.1.2](https://github.com/favware/node-packages/compare/@favware/zalgo@1.1.1...@favware/zalgo@1.1.2) (2019-10-13)

### Bug Fixes

- completely re-do tsconfig standardized ([9eb030f](https://github.com/favware/node-packages/commit/9eb030fdf1deb75d5ae8b273d0e9c359bcb985a1))

## [1.1.1](https://github.com/favware/node-packages/compare/@favware/zalgo@1.1.0...@favware/zalgo@1.1.1) (2019-10-13)

**Note:** Version bump only for package @favware/zalgo

# [1.1.0](https://github.com/favware/node-packages/compare/@favware/zalgo@1.0.6...@favware/zalgo@1.1.0) (2019-10-13)

### Features

- standardize TS configuration ([47696f4](https://github.com/favware/node-packages/commit/47696f4e1dd2632b305ff9789cdd6c473fa709ca))

## [1.0.6](https://github.com/favware/node-packages/compare/@favware/zalgo@1.0.5...@favware/zalgo@1.0.6) (2019-09-29)

### Bug Fixes

- **zalgo:** remove unnecessary peer dependencies ([74e6af6](https://github.com/favware/node-packages/commit/74e6af6))

## [1.0.5](https://github.com/favware/node-packages/compare/@favware/zalgo@1.0.4...@favware/zalgo@1.0.5) (2019-09-11)

### Bug Fixes

- **zalgo:** fix prop in obj not being valid ([807be90](https://github.com/favware/node-packages/commit/807be90))

## [1.0.4](https://github.com/favware/node-packages/compare/@favware/zalgo@1.0.3...@favware/zalgo@1.0.4) (2019-09-11)

**Note:** Version bump only for package @favware/zalgo

## [1.0.3](https://github.com/favware/node-packages/compare/@favware/zalgo@1.0.2...@favware/zalgo@1.0.3) (2019-08-21)

**Note:** Version bump only for package @favware/zalgo

## 1.0.2 (2019-08-18)

### Reverts

- "ci: another attempt to fix CI config" ([37c385f](https://github.com/favware/node-packages/commit/37c385f))
