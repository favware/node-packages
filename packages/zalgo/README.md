<div align="center">
  <p>
  <a href="https://favware.tech/zalgo"><img src="https://storage.googleapis.com/data-sunlight-146313.appspot.com/website-project-icons/zalgo.png" height="200" alt="logo"/></a>
  </p>

  <p>
<h1> @favware/zalgo</h1>
<h3> Invoke and banish the power of the Zalgo Monster</h3>
  </p>
</div>

---

**Project Status**

[![GitHub](https://img.shields.io/github/license/favware/node-packages?logo=github&style=flat-square)](https://github.com/favware/node-packages/blob/master/LICENSE.md)
[![Codecov](https://img.shields.io/codecov/c/github/favware/node-packages?logo=codecov&style=flat-square)](https://codecov.io/gh/favware/node-packages)

**Bundle Sizes**

[![npm bundle size](https://img.shields.io/bundlephobia/min/@favware/zalgo?label=zalgo%20-%20minified&logo=webpack&style=flat-square)](https://bundlephobia.com/result?p=@favware/zalgo)
[![npm bundle size](https://img.shields.io/bundlephobia/minzip/@favware/zalgo?label=zalgo%20-%20minzipped&logo=webpack&style=flat-square)](https://bundlephobia.com/result?p=@favware/zalgo)
[![npm](https://img.shields.io/npm/v/@favware/zalgo?color=crimson&label=zalgo%20version&logo=npm&style=flat-square)](https://www.npmjs.com/package/@favware/zalgo)

**Social Media and Donations**

[![Join Discord server](https://img.shields.io/discord/512303595966824458?color=697EC4&label=Join%20Discord%20Server&logo=discord&logoColor=FDFEFE&style=flat-square)](https://favware.tech/redirect/server)
[![Twitter Follow](https://img.shields.io/twitter/follow/favna_?label=Follow%20@Favna_&logo=twitter&colorB=1DA1F2&style=flat-square)](https://twitter.com/Favna_/follow)
[![Patreon Donate](https://img.shields.io/badge/patreon-donate-brightgreen.svg?label=Donate%20with%20Patreon&logo=patreon&colorB=F96854&style=flat-square&link=https://www.patreon.com/bePatron?u=9336537)](https://www.patreon.com/bePatron?u=9336537)
[![PayPal Donate](https://img.shields.io/badge/paypal-donate-brightgreen.svg?label=Donate%20with%20Paypal&logo=paypal&colorB=00457C&style=flat-square&link=https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=XMAYCF9SDHZ34)](https://www.patreon.com/bePatron?u=9336537)

---

This library uses the Zalgo character set from [Acme::Zalgo](https://metacpan.org/pod/Acme::Zalgo)

# Install

Install with [yarn](https://yarnpkg.com) or [npm](https://www.npmjs.com/):

```sh
yarn add @favware/zalgo

# npm install @favware/zalgo
```

---

# Usage

```js
const { zalgo, banish, chars } = require('awesome-zalgo');
// Or for babel / typescript:
// import zalgo, { banish, chars } from 'awesome-zalgo';

// Zalgo usage
zalgo('some text');

// > ŝ̜̩͇̼̥̼́̏͢o͎͊͜ḿ̛̩̳̖͕̞̩̭ͪe͖̺̣̹̺̋̀͛̽͝ ̖͍̭͓̯̠͑͑͢t̼̪̋͌͢eͯ̋͏͖͎͍̩̭̮x̢͚̄̾̀̈ͧ̓ͩ̚t̪ͫ͝

// Banish usage
banish('ŝ̜̩͇̼̥̼́̏͢o͎͊͜ḿ̛̩̳̖͕̞̩̭ͪe͖̺̣̹̺̋̀͛̽͝ ̖͍̭͓̯̠͑͑͢t̼̪̋͌͢eͯ̋͏͖͎͍̩̭̮x̢͚̄̾̀̈ͧ̓ͩ̚t̪ͫ͝');

// > some text

// To log all the supported chars use:
console.log(chars);
```

## [API Documentation](https://favware.github.io/node-packages/modules/_favware_zalgo.html)
