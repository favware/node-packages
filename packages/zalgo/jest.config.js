module.exports = {
  displayName: 'ts-jest',
  preset: 'ts-jest',
  testEnvironment: 'node',
  testRunner: 'jest-circus/runner'
};
