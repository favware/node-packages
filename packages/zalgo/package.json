{
  "name": "@favware/zalgo",
  "version": "2.0.0",
  "description": "Invoke and banish the power of the Zalgo Monster",
  "author": "@favware",
  "license": "MIT",
  "main": "dist/index.js",
  "module": "dist/index.es.js",
  "types": "dist/index.d.ts",
  "homepage": "https://github.com/favware/node-packages/tree/master/packages/zalgo#readme",
  "scripts": {
    "test": "jest",
    "lint": "eslint src __tests__ --ext ts --fix -c ../../.eslintrc",
    "prebuild": "yarn lint && yarn test",
    "build": "tsc -b src",
    "start": "yarn build -w"
  },
  "files": [
    "dist",
    "!dist/.tsbuildinfo"
  ],
  "engines": {
    "node": ">=10",
    "npm": ">=6"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/favware/node-packages.git",
    "directory": "packages/zalgo"
  },
  "bugs": {
    "url": "https://github.com/favware/node-packages/issues"
  },
  "keywords": [
    "zalgo",
    "banish",
    "mess up",
    "favware",
    "typescript",
    "ts"
  ],
  "directories": {
    "lib": "src",
    "test": "__tests__"
  },
  "publishConfig": {
    "access": "public"
  }
}
