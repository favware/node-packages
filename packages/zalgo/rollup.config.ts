import { resolve as resolveDir } from 'path';
import cleaner from 'rollup-plugin-cleaner';
import { terser } from 'rollup-plugin-terser';
import typescript from 'rollup-plugin-typescript2';

export default {
  input: 'src/index.ts',
  output: [
    {
      file: './dist/index.js',
      format: 'cjs',
      exports: 'named',
      sourcemap: true
    },
    {
      file: './dist/index.mjs',
      format: 'es',
      exports: 'named',
      sourcemap: true
    },
    {
      file: './dist/index.umd.js',
      format: 'umd',
      exports: 'named',
      name: 'FavwareZalgo',
      sourcemap: true
    }
  ],
  plugins: [
    cleaner({
      targets: ['./dist/']
    }),
    typescript({ tsconfig: resolveDir(__dirname, 'src', 'tsconfig.json') }),
    terser({
      ecma: 2019,
      // This will ensure that whenever Rollup is in watch (dev) mode, console logs will not be removed
      compress: { drop_console: !Reflect.has(process.env, 'ROLLUP_WATCH') },
      format: { comments: false },
      keep_classnames: true,
      keep_fnames: true
    })
  ]
};
