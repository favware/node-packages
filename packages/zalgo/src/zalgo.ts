import { chars } from './chars';
import type { ZalgoCountsMap, ZalgoOptions } from './interfaces';
import { randomizer, unicodeStringSplitter } from './utils';

/**
 * Zalgofies any given text
 *
 * @param text Input text to zalgolize
 * @param options Options for the Zalgo
 * @returns The door to hell
 * @example
 * ```ts
 * zalgo('some text') // ==> ŝ̜̩͇̼̥̼́̏͢o͎͊͜ḿ̛̩̳̖͕̞̩̭ͪe͖̺̣̹̺̋̀͛̽͝ ̖͍̭͓̯̠͑͑͢t̼̪̋͌͢eͯ̋͏͖͎͍̩̭̮x̢͚̄̾̀̈ͧ̓ͩ̚t̪ͫ͝
 *```
 */
export const zalgo = (
  text: string,
  options: ZalgoOptions = {
    up: true,
    middle: true,
    down: true,
    size: ''
  }
): string => {
  try {
    if (!text) throw new Error('no_input');
    if (typeof text !== 'string') throw new Error('not_a_string');
    if (!Reflect.has(options, 'up')) options.up = true;
    if (!Reflect.has(options, 'middle')) options.middle = true;
    if (!Reflect.has(options, 'down')) options.down = true;
    if (!Reflect.has(options, 'size')) options.size = '';

    const splitText = unicodeStringSplitter(text);
    const splitTextLength = splitText.length;

    let counts: ZalgoCountsMap;
    let result = '';
    const types: string[] = [];

    if (options.up) types.push('up');
    if (options.middle) types.push('middle');
    if (options.down) types.push('down');

    for (let i = 0; i < splitTextLength; i++) {
      if (chars.pattern!.test(splitText[i])) continue;

      // Skip Emojis
      if (splitText[i].length > 1) {
        result += splitText[i];
        continue;
      }

      counts = { up: 0, middle: 0, down: 0 };

      if (options.size === 'mini') counts = { up: randomizer(8), middle: randomizer(2), down: randomizer(8) };
      else if (options.size === 'maxi') counts = { up: randomizer(16) + 3, middle: randomizer(4) + 1, down: randomizer(64) + 3 };
      else counts = { up: randomizer(8) + 1, middle: randomizer(3), down: randomizer(8) + 1 };

      result += text[i];

      for (let j = 0, m = types.length; j < m; j++) {
        const type = types[j];
        let count = counts[type];
        const tchars = chars[type] as string[];
        const max = tchars.length - 1;

        while (count--) {
          result += tchars[randomizer(max)];
        }
      }
    }

    return result;
  } catch (err) {
    if (/(?:no_input)/i.test((err as Error).message)) throw SyntaxError('The zalgo function at least requires some text as input!');
    if (/(?:not_a_string)/i.test((err as Error).message)) throw new SyntaxError('The zalgo function expects input of type string as first argument!');
    throw err;
  }
};

/**
 * De-zalgolize any text
 * @param purgeable Text to remove zalgo from
 * @returns The door to heaven
 * @example
 * ```ts
 * banish('ŝ̜̩͇̼̥̼́̏͢o͎͊͜ḿ̛̩̳̖͕̞̩̭ͪe͖̺̣̹̺̋̀͛̽͝ ̖͍̭͓̯̠͑͑͢t̼̪̋͌͢eͯ̋͏͖͎͍̩̭̮x̢͚̄̾̀̈ͧ̓ͩ̚t̪ͫ͝') // ==> some text
 * ```
 */
export const banish = (purgeable: string): string => {
  try {
    if (!purgeable) throw new Error('no_input');
    if (typeof purgeable !== 'string') throw new Error('not_a_string');

    return purgeable.replace(chars.pattern!, '');
  } catch (err) {
    if (/(?:no_input)/i.test((err as Error).message)) throw new SyntaxError('The banish function at least requires some text as input!');
    if (/(?:not_a_string)/i.test((err as Error).message))
      throw new SyntaxError('The banish function expects input of type string as first argument!');
    throw err;
  }
};
