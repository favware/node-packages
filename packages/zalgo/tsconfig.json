{
  "extends": "../ts-config/tsconfig.json",
  "compilerOptions": {
    "outDir": "build",
    "rootDir": "src"
  },
  "include": ["src"],
  "exclude": ["node_modules", "**/*.test.ts"]
}
