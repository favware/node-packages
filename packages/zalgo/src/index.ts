import { zalgo } from './zalgo';

export { chars } from './chars';
export * from './interfaces';
export { banish, zalgo } from './zalgo';

export default zalgo;
