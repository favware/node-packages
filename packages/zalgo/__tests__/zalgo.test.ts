import zalgo, { chars, banish, zalgo as NamedZalgo } from '../src';

describe('Snapshot Testing', () => {
  test('zalgo should match snapshot', () => {
    expect(zalgo).toMatchSnapshot();
  });

  test('chars should match snapshot', () => {
    expect(chars).toMatchSnapshot();
  });

  test('banish should match snapshot', () => {
    expect(banish).toMatchSnapshot();
  });

  test('NamedZalgo should match snapshot', () => {
    expect(NamedZalgo).toMatchSnapshot();
  });
});

describe('No option testing', () => {
  test('should zalgo normally', () => {
    const text = 'That rug really tied the room together';
    const evil = zalgo(text);

    expect(chars.pattern!.test(evil)).toBe(true);
  });

  test('should skip emojis', () => {
    const text = 'These are some annoying emojis: ðŸ’¦ðŸ˜‚ðŸ”¥';
    const evil = zalgo(text);

    expect(chars.pattern!.test(evil)).toBe(true);
  });
});

describe('Option Testing', () => {
  test('should support mini zalgo', () => {
    const text = 'the answer to life, the universe and everything is 42';
    const evil = zalgo(text, { size: 'mini' });

    expect(chars.pattern!.test(evil)).toBe(true);
  });

  test('should support maxi zalgo', () => {
    const text = "I'm sorry, Dave. I'm afraid I can't do that.";
    const evil = zalgo(text, { size: 'maxi' });

    expect(chars.pattern!.test(evil)).toBe(true);
  });

  test('should support only upper zalgo', () => {
    const text = 'With great power there must also come â€“ great responsibility.';
    const evil = zalgo(text, { up: true, middle: false, down: true });

    expect(chars.pattern!.test(evil)).toBe(true);
  });

  test('should support only middle zalgo', () => {
    const text = 'So Long, and Thanks for All the Fish';
    const evil = zalgo(text, { up: false, middle: true, down: false });

    expect(chars.pattern!.test(evil)).toBe(true);
  });

  test('should support only down zalgo', () => {
    const text = 'I know kung fu.';
    const evil = zalgo(text, { up: false, middle: false, down: true });

    expect(chars.pattern!.test(evil)).toBe(true);
  });
});
