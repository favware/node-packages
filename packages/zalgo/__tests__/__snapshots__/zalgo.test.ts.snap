// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing NamedZalgo should match snapshot 1`] = `[Function]`;

exports[`Snapshot Testing banish should match snapshot 1`] = `[Function]`;

exports[`Snapshot Testing chars should match snapshot 1`] = `
Object {
  "all": Array [
    "̍",
    "̎",
    "̄",
    "̅",
    "̿",
    "̑",
    "̆",
    "̐",
    "͒",
    "͗",
    "͑",
    "̇",
    "̈",
    "̊",
    "͂",
    "̓",
    "̈́",
    "͊",
    "͋",
    "͌",
    "̃",
    "̂",
    "̌",
    "͐",
    "̀",
    "́",
    "̋",
    "̏",
    "̒",
    "̓",
    "̔",
    "̽",
    "̉",
    "ͣ",
    "ͤ",
    "ͥ",
    "ͦ",
    "ͧ",
    "ͨ",
    "ͩ",
    "ͪ",
    "ͫ",
    "ͬ",
    "ͭ",
    "ͮ",
    "ͯ",
    "̾",
    "͛",
    "͆",
    "̚",
    "̕",
    "̛",
    "̀",
    "́",
    "͘",
    "̡",
    "̢",
    "̧",
    "̨",
    "̴",
    "̵",
    "̶",
    "͏",
    "͜",
    "͝",
    "͞",
    "͟",
    "͠",
    "͢",
    "̸",
    "̷",
    "͡",
    "҉",
    "̖",
    "̗",
    "̘",
    "̙",
    "̜",
    "̝",
    "̞",
    "̟",
    "̠",
    "̤",
    "̥",
    "̦",
    "̩",
    "̪",
    "̫",
    "̬",
    "̭",
    "̮",
    "̯",
    "̰",
    "̱",
    "̲",
    "̳",
    "̹",
    "̺",
    "̻",
    "̼",
    "ͅ",
    "͇",
    "͈",
    "͉",
    "͍",
    "͎",
    "͓",
    "͔",
    "͕",
    "͖",
    "͙",
    "͚",
    "̣",
  ],
  "down": Array [
    "̖",
    "̗",
    "̘",
    "̙",
    "̜",
    "̝",
    "̞",
    "̟",
    "̠",
    "̤",
    "̥",
    "̦",
    "̩",
    "̪",
    "̫",
    "̬",
    "̭",
    "̮",
    "̯",
    "̰",
    "̱",
    "̲",
    "̳",
    "̹",
    "̺",
    "̻",
    "̼",
    "ͅ",
    "͇",
    "͈",
    "͉",
    "͍",
    "͎",
    "͓",
    "͔",
    "͕",
    "͖",
    "͙",
    "͚",
    "̣",
  ],
  "middle": Array [
    "̕",
    "̛",
    "̀",
    "́",
    "͘",
    "̡",
    "̢",
    "̧",
    "̨",
    "̴",
    "̵",
    "̶",
    "͏",
    "͜",
    "͝",
    "͞",
    "͟",
    "͠",
    "͢",
    "̸",
    "̷",
    "͡",
    "҉",
  ],
  "pattern": /\\(̍\\|̎\\|̄\\|̅\\|̿\\|̑\\|̆\\|̐\\|͒\\|͗\\|͑\\|̇\\|̈\\|̊\\|͂\\|̓\\|̈́\\|͊\\|͋\\|͌\\|̃\\|̂\\|̌\\|͐\\|̀\\|́\\|̋\\|̏\\|̒\\|̓\\|̔\\|̽\\|̉\\|ͣ\\|ͤ\\|ͥ\\|ͦ\\|ͧ\\|ͨ\\|ͩ\\|ͪ\\|ͫ\\|ͬ\\|ͭ\\|ͮ\\|ͯ\\|̾\\|͛\\|͆\\|̚\\|̕\\|̛\\|̀\\|́\\|͘\\|̡\\|̢\\|̧\\|̨\\|̴\\|̵\\|̶\\|͏\\|͜\\|͝\\|͞\\|͟\\|͠\\|͢\\|̸\\|̷\\|͡\\|҉\\|̖\\|̗\\|̘\\|̙\\|̜\\|̝\\|̞\\|̟\\|̠\\|̤\\|̥\\|̦\\|̩\\|̪\\|̫\\|̬\\|̭\\|̮\\|̯\\|̰\\|̱\\|̲\\|̳\\|̹\\|̺\\|̻\\|̼\\|ͅ\\|͇\\|͈\\|͉\\|͍\\|͎\\|͓\\|͔\\|͕\\|͖\\|͙\\|͚\\|̣\\)/g,
  "up": Array [
    "̍",
    "̎",
    "̄",
    "̅",
    "̿",
    "̑",
    "̆",
    "̐",
    "͒",
    "͗",
    "͑",
    "̇",
    "̈",
    "̊",
    "͂",
    "̓",
    "̈́",
    "͊",
    "͋",
    "͌",
    "̃",
    "̂",
    "̌",
    "͐",
    "̀",
    "́",
    "̋",
    "̏",
    "̒",
    "̓",
    "̔",
    "̽",
    "̉",
    "ͣ",
    "ͤ",
    "ͥ",
    "ͦ",
    "ͧ",
    "ͨ",
    "ͩ",
    "ͪ",
    "ͫ",
    "ͬ",
    "ͭ",
    "ͮ",
    "ͯ",
    "̾",
    "͛",
    "͆",
    "̚",
  ],
}
`;

exports[`Snapshot Testing zalgo should match snapshot 1`] = `[Function]`;
